apps/oc_chef_wm/src/oc_chef_wm_named_cookbook_artifact_version.erl

```diff
@@ -0,0 +1,153 @@
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil; fill-column: 92 -*-
+%% ex: ts=4 sw=4 et
+%% @author Jean Rouge <jean@chef.io>
+%% Copyright 2013-2015 Chef Software, Inc. All Rights Reserved.
+
+-module(oc_chef_wm_named_cookbook_artifact_version).
+
+-include("../../include/oc_chef_wm.hrl").
+
+-behaviour(chef_wm).
+
+%% Webmachine resource callbacks
+-mixin([{oc_chef_wm_base, [content_types_accepted/2,
+                           content_types_provided/2,
+                           finish_request/2,
+                           malformed_request/2,
+                           ping/2,
+                           forbidden/2,
+                           is_authorized/2,
+                           service_available/2]}]).
+
+-export([validate_request/3,
+         auth_info/2,
+         init_resource_state/1,
+         init/1,
+         request_type/0,
+         allowed_methods/2,
+         post_is_create/2,
+         from_json/2,
+         resource_exists/2,
+         to_json/2,
+         create_path/2,
+         malformed_request_message/3,
+         conflict_message/1]).
+
+init(Config) ->
+    oc_chef_wm_base:init(?MODULE, Config).
+
+init_resource_state(_Config) ->
+    {ok, #cookbook_artifact_version_state{}}.
+
+request_type() ->
+    "cookbook_artifact_versions".
+
+allowed_methods(Req, State) ->
+    {['GET', 'POST'], Req, State}.
+
+post_is_create(Req, State) ->
+    {true, Req, State}.
+
+create_path(Req, State) ->
+    Name = wrq:path_info(name, Req),
+    {Name, Req, State}.
+
+resource_exists(Req, State) ->
+    {true, Req, State}.
+
+validate_request('GET', Req, State) ->
+    valid_request(Req, State);
+validate_request('POST', Req, #base_state{resource_state = CAVState} = State) ->
+    CAVData = validate_json(Req),
+    NewResourceState = CAVState#cookbook_artifact_version_state{cookbook_artifact_version_data = CAVData},
+    valid_request(Req, State#base_state{resource_state = NewResourceState}).
+
+validate_json(Req) ->
+    Body = wrq:req_body(Req),
+    Ejson = chef_json:decode_body(Body),
+    [validate_url_param(Req, Ejson, Param) || Param <- [name, identifier]],
+    case oc_chef_cookbook_artifact_version:validate_json(Ejson) of
+        ok -> Ejson;
+        Invalid -> throw(Invalid)
+    end.
+
+validate_url_param(Req, Ejson, Param) ->
+    FromReq = erlang:list_to_binary(wrq:path_info(Param, Req)),
+    ParamBin = erlang:atom_to_binary(Param, utf8),
+    FromEjson = ej:get({ParamBin}, Ejson),
+    case FromReq =:= FromEjson of
+        true ->
+            ok;
+        false ->
+            erlang:throw({mismatch, {ParamBin, FromEjson, FromReq}})
```

Comment:
Dan's point is entirely valid.

Also, it's partly because we want this to be present in the serialized object we save. But then we could be adding it dynamically.